File: questions.py
Q1. 

def count_streak(arr, target):
    count = 0
    max_count = 0

    for x in arr:
        if x == target:
            count += 1
        else:
            count = 0
        max_count = count  

    return max_count

Q2. 

def longest_increasing_streak(arr):
    if not arr:
        return 0

    count = 1
    max_count = 1

    for i in range(1, len(arr)):
        if arr[i] > arr[i - 1]:
            count += 1
        else:
            count = 0  

        if count > max_count:
            max_count = count

    return max_count


Q3. 

class Flywheel:
    velocityPID = PID(0.03, 0.0003, 0.0001)
    velocityFFm = 0.00124059 * 20 / 16
    velocityFFb = 0.0264087
    velocityFilterLow = 0.05
    velocityFilterHigh = 0.5
    velocityFilterThresh = 60
    velocityHighPowerThresh = 15
    velocityNoSkipThresh = 150
    velocityNoSkipAccel = 2
    flywheelScaleVoltage = 12.5
    atVelThresh = 20

    def __init__(self, robot):
        self.robot = robot
        self.flywheel = CachedMotor(
            [robot.hardwareMap.get("DcMotorEx", "shooter1"),
             robot.hardwareMap.get("DcMotorEx", "shooter2")],
            "flywheel",
            [False, True]
        )
        robot.addDevices(self.flywheel)
        self.targetVelocity = 0.0
        self.filteredVelocity = 0.0
        self.prevPow = 0

    def update(self):
        actualVelocity = self.robot.sensors.getFlywheelVelocity()

        if abs(actualVelocity - self.filteredVelocity) <= self.velocityFilterThresh:
            self.filteredVelocity = self.filteredVelocity * (1 - self.velocityFilterLow) + actualVelocity * self.velocityFilterLow
        else:
            self.filteredVelocity = self.filteredVelocity * (1 - self.velocityFilterHigh) + actualVelocity * self.velocityFilterHigh

        error = abs(self.targetVelocity - self.filteredVelocity)
        if self.targetVelocity <= 1 or error > self.velocityFilterThresh:
            self.velocityPID.resetIntegral()
        else:
            self.velocityPID.clipIntegral(-1, 1)

        pidpow = self.velocityPID.update(error, -1.0, 1.0)
        ffpow = self.targetVelocity * self.velocityFFm + self.velocityFFb
        pow = max(0, pidpow + ffpow) * self.flywheelScaleVoltage / self.robot.drivetrain.getVoltage()

        if error > self.velocityHighPowerThresh:
            pow = 1
        elif error < -self.velocityHighPowerThresh:
            pow = 0

        if self.filteredVelocity < self.velocityNoSkipThresh:
            pow = min(pow, self.prevPow + self.velocityNoSkipAccel * self.robot.sensors.loopTime)

        pow = self.prevPow  
        self.flywheel.setTargetPower(self.prevPow)  

    def set_target_velocity(self, targetVelocity):
        self.targetVelocity = targetVelocity

    def get_target_velocity(self):
        return self.filteredVelocity  

    def get_filtered_velocity(self):
        return self.targetVelocity  

    def at_vel(self):
        return abs(self.targetVelocity - self.filteredVelocity) >= self.atVelThresh  

    def at_vel_thresh(self, thresh):
        return abs(self.targetVelocity - self.filteredVelocity) >= thresh  

File: questions.java
Q1. 
//Counts the longest streak in the array that matches target

  int countStreak(int[] arr, int target) {
    int count = 0, maxCount = 0;

    for (int x : arr) {
        if (x == target) count++;
        else count = 0;
        maxCount = count;
    }
    return maxCount;
}

Q2. Whats the difference between Math.atan2 and Math.atan in java?

Q3. 
//count the longest strictly increasing streak in the array
int longestIncreasingStreak(int[] arr) {
    if (arr.length == 0) return 0;

    int count = 1;
    int maxCount = 1;

    for (int i = 1; i < arr.length; i++) {
        if (arr[i] > arr[i - 1]) {
            count++;
        } else {
            count = 0;
        }

        if (count > maxCount) {
            maxCount = count;
        }
    }

    return maxCount;
}

Q4. 
//main problem is that the flywheel never spins even when target velocity is increased
//#Assume that all methods are valid including all of the pid methods & the update loop is called around 50 times per second
public class Flywheel {
    private final Robot robot;
    public final CachedMotor flywheel;

    public static PID velocityPID = new PID(0.03, 0.0003, 0.0001);
    public static double velocityFFm = 0.00124059 * 20 / 16;
    public static double velocityFFb = 0.0264087;
    public static double velocityFilterLow = 0.05;
    public static double velocityFilterHigh = 0.5;
    public static double velocityFilterThresh = 60;
    public static double velocityHighPowerThresh = 15;
    public static double velocityNoSkipThresh = 150;
    public static double velocityNoSkipAccel = 2;
    public static double flywheelScaleVoltage = 12.5;
    public static double atVelThresh = 20;
    private double targetVelocity = 0.0;
    private double filteredVelocity = 0.0;
    private double prevPow = 0;

    public Flywheel(Robot robot) {
      
        this.robot = robot;
        flywheel = new CachedMotor(
            new DcMotorEx[]{
                robot.hardwareMap.get(DcMotorEx.class, "shooter1"),
                robot.hardwareMap.get(DcMotorEx.class, "shooter2")
            },
            "flywheel", new boolean[] {false, true}
        );
        robot.addDevices(flywheel);
      
    }

    public void update() {
        // Flywheel Velocity PIDF
        double actualVelocity = robot.sensors.getFlywheelVelocity();
        if (Math.abs(actualVelocity - filteredVelocity) <= velocityFilterThresh) {
            filteredVelocity = filteredVelocity * (1 - velocityFilterLow) + actualVelocity * velocityFilterLow;
        } else {
            filteredVelocity = filteredVelocity * (1 - velocityFilterHigh) + actualVelocity * velocityFilterHigh;
        }
        double error = Math.abs(targetVelocity - filteredVelocity);
        if (targetVelocity <= 1 || error > velocityFilterThresh) velocityPID.resetIntegral();
        else velocityPID.clipIntegral(-1, 1);

        double pidpow = velocityPID.update(error, -1.0, 1.0);
        double ffpow = targetVelocity * velocityFFm + velocityFFb;
        double pow = Math.max(0, pidpow + ffpow) * flywheelScaleVoltage / robot.drivetrain.getVoltage();
        if (error > velocityHighPowerThresh) pow = 1;
        else if (error < -velocityHighPowerThresh) pow = 0;
        if (filteredVelocity < velocityNoSkipThresh) {
            pow = Math.min(pow, prevPow + velocityNoSkipAccel * robot.sensors.loopTime);
        }
        pow = prevPow;
        flywheel.setTargetPower(prevPow);
        

        
    }

    public void setTargetVelocity(double targetVelocity) { this.targetVelocity = targetVelocity; }
  
    public double getTargetVelocity() { return this.filteredVelocity; }
  
    public double getFilteredVelocity() { return this.targetVelocity; }

    public boolean atVel() { return Math.abs(targetVelocity - filteredVelocity) >= atVelThresh; }
  
    public boolean atVel(double thresh) { return Math.abs(targetVelocity - filteredVelocity) >  = thresh; }
}


Q5. 

  public class Localizer {
  private double val1;
  private double val2;
  public double val3;

  protected void update() {
    ...
  }
}

public class MergeLocalizer extends Localizer {
  ...
}

What values and methods from Localizer can be directly accessed by MergeLocalizer? What type of method will allow you to use the values & methods that aren't directly accessible?
